### PR TITLE
add ur analytic ik solver

### PR DESCRIPTION
**Pull request by matafela**
# Description

Add ur analytic ik solver.

TODO:

- [x] benchmark
- [x] unittest

benchmark:

## Time & Memory

| sample_size | impl | component | cost_time_ms | cpu_delta_mb | gpu_delta_mb | peak_gpu_mb |
| --- | --- | --- | --- | --- | --- | --- |
| 100 | ur_cpu | ur_ik | 0.402335 | 4.492188 | 0.000000 | 8.159180 |
| 100 | ur_cuda | ur_ik | 0.727567 | 301.230469 | 0.002930 | 8.200684 |
| 1000 | ur_cpu | ur_ik | 2.853392 | 0.000000 | 0.000000 | 8.179199 |
| 1000 | ur_cuda | ur_ik | 0.455092 | 0.003906 | 0.023926 | 8.561523 |
| 10000 | ur_cpu | ur_ik | 25.540989 | 0.003906 | 0.000000 | 8.343750 |
| 10000 | ur_cuda | ur_ik | 0.418456 | 0.011719 | 0.238770 | 12.161621 |

## Success & Other Metrics

| sample_size | impl | component | success_rate | translation_err_mm | rotation_err_deg |
| --- | --- | --- | --- | --- | --- |
| 100 | ur_cpu | ur_ik | 1.000000 | 0.000148 | 0.016093 |
| 100 | ur_cuda | ur_ik | 1.000000 | 0.000159 | 0.004525 |
| 1000 | ur_cpu | ur_ik | 1.000000 | 0.000159 | 0.016271 |
| 1000 | ur_cuda | ur_ik | 1.000000 | 0.000166 | 0.005868 |
| 10000 | ur_cpu | ur_ik | 1.000000 | 0.000159 | 0.016187 |
| 10000 | ur_cuda | ur_ik | 1.000000 | 0.000170 | 0.005460 |

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have run the `black .` command to format the code base.
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Dependencies have been updated, if applicable.

### CODE REVIEW

File: embodichain/lab/sim/solvers/ur_solver.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
+import torch
```
Comment (yuecideng) [inline embodichain/lab/sim/solvers/ur_solver.py:17]: Add copyright

File: embodichain/lab/sim/solvers/ur_solver.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+import numpy as np
```
Comment (yuecideng) [inline embodichain/lab/sim/solvers/ur_solver.py:18]: Add corresponding docs for UR Solver

File: embodichain/lab/sim/solvers/ur_solver.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            .to(device=device)
+        )
+
+        if return_all_solutions:
+            return all_solutions_validity, all_solutions
+        # Select ik qpos based on the closest distance to the seed qpos
+        qpos_seed_expanded = qpos_seed.unsqueeze(1).expand(-1, N_SOL, -1)
+        distances = torch.norm(all_solutions - qpos_seed_expanded, dim=-1)
+        # fill invalid solutions with inf distance
+        distances[~all_solutions_validity] = float("inf")
```
Comment (yuecideng) [inline embodichain/lab/sim/solvers/ur_solver.py:210]: This selection masks only the FK-valid candidates, but it never filters candidates against `lower_qpos_limits` / `upper_qpos_limits`. That means `compute_ik` can return `success=True` for joints outside the robot or user-provided limits. I was able to reproduce this with `user_qpos_limits=[[-0.1] * 6, [0.1] * 6]`: the returned solution had several joints outside that range while validity stayed true. Please map periodic solutions into the allowed range where possible, then mark out-of-range candidates invalid before nearest-solution selection.

File: embodichain/utils/warp/kinematics/ur_solver.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    # Write solutions and perform FK-based validity check.
+    for j in range(N_SOL):
+        qpos_start = i * DOF * N_SOL + j * DOF
+        q1 = theta[j * DOF + 0]
+        q2 = theta[j * DOF + 1]
+        q3 = theta[j * DOF + 2]
+        q4 = theta[j * DOF + 3]
+        q5 = theta[j * DOF + 4]
+        q6 = theta[j * DOF + 5]
+        qpos[qpos_start + 0] = normalize_to_pi(q1)
```
Comment (yuecideng) [inline embodichain/utils/warp/kinematics/ur_solver.py]: Normalizing every solution to `[-pi, pi]` loses valid periodic joint representations. The UR10 URDF allows roughly `[-2pi, 2pi]`, so a seed near `1.5pi` can be forced to the wrapped `-0.5pi` branch and the nearest-solution logic may create a discontinuous jump. It would be better to generate/wrap candidates relative to the configured joint limits and seed before selecting the nearest solution.

File: tests/sim/solvers/test_ur_solver.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            ],
+            init_pos=(0, 0, 0),
+        )
+        self.robot: Robot = self.sim.add_robot(cfg=cfg)
+
+    def test_ik(self):
+        # Test inverse kinematics (IK) with a 1x4x4 homogeneous matrix pose and a joint_seed
+        arm_name = "arm"
+        # qpos_limit = self.robot.get_qpos_limits(name=arm_name)
+        qpos_limit = torch.tensor(
```
Comment (yuecideng) [inline tests/sim/solvers/test_ur_solver.py]: This hardcoded `[-pi, pi]` range hides two important behaviors: the actual URDF limits are wider, and the solver currently normalizes outputs into `[-pi, pi]`. Please use `robot.get_qpos_limits(name=arm_name)` for at least one coverage path, and add targeted cases for custom `user_qpos_limits`, `qpos_seed=None`, and a 1-D seed so the public solver contract is exercised.

File: embodichain/lab/sim/solvers/ur_solver.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    a3: float = -0.5723
+    d4: float = 0.163941
+    d5: float = 0.1157
+    d6: float = 0.0922
+
+    alpha1: float = torch.pi * 0.5
+    alpha4: float = torch.pi * 0.5
+    alpha5: float = -torch.pi * 0.5
+
+    def __post_init__(self):
```
Comment (yuecideng) [inline embodichain/lab/sim/solvers/ur_solver.py:49]: `__post_init__` overwrites `d1/a2/a3/d4/d5/d6` and `urdf_path` unconditionally from `ur_type`. Since these fields are public config values, a caller cannot provide calibrated DH parameters or a custom URDF while still selecting a UR family. Either only fill defaults when the user did not override them, or make the solver type presets explicit and avoid exposing these as independently configurable values.

**Review by yuecideng — COMMENTED** [review]
Adding a few comments on correctness issues that are not covered by the existing automated review.

File: embodichain/lab/sim/solvers/ur_solver.py (hunk 6 above)
Comment (matafela) [reply embodichain/lab/sim/solvers/ur_solver.py:49]: Currently still use hard coded ur DH parameter. Latter we will have a new URCfg as a subclass of RobotCfg

**Review by yuecideng — APPROVED** [review]
